cyper-core/src/stream.rs

```diff
@@ -179,36 +228,38 @@ macro_rules! poll_future_would_block {
 
 #[cfg(not(feature = "read_buf"))]
 #[inline]
-fn read_buf(reader: &mut impl io::Read, buf: &mut tokio::io::ReadBuf<'_>) -> io::Result<()> {
-    let slice = buf.initialize_unfilled();
+fn read_buf(reader: &mut impl io::Read, mut buf: hyper::rt::ReadBufCursor<'_>) -> io::Result<()> {
+    use std::mem::MaybeUninit;
+
+    let slice = unsafe { buf.as_mut() };
+    slice.fill(MaybeUninit::new(0));
+    let slice = unsafe { &mut *(slice as *mut _ as *mut [u8]) };
     let len = reader.read(slice)?;
-    buf.advance(len);
+    unsafe { buf.advance(len) };
     Ok(())
 }
 
 #[cfg(feature = "read_buf")]
 #[inline]
-fn read_buf(reader: &mut impl io::Read, buf: &mut tokio::io::ReadBuf<'_>) -> io::Result<()> {
-    let slice = unsafe { buf.unfilled_mut() };
+fn read_buf(reader: &mut impl io::Read, mut buf: hyper::rt::ReadBufCursor<'_>) -> io::Result<()> {
+    let slice = unsafe { buf.as_mut() };
     let len = {
         let mut borrowed_buf = io::BorrowedBuf::from(slice);
         let mut cursor = borrowed_buf.unfilled();
         reader.read_buf(cursor.reborrow())?;
         cursor.written()
     };
-    unsafe { buf.assume_init(len) };
-    buf.advance(len);
+    unsafe { buf.advance(len) };
     Ok(())
 }
 
-impl tokio::io::AsyncRead for HttpStream {
+impl<S: AsyncRead + Unpin + 'static> hyper::rt::Read for HyperStream<S> {
     fn poll_read(
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-        buf: &mut tokio::io::ReadBuf<'_>,
+        buf: hyper::rt::ReadBufCursor<'_>,
     ) -> Poll<io::Result<()>> {
-        let inner: &'static mut SyncStream<HttpStreamInner> =
-            unsafe { &mut *(self.inner.deref_mut() as *mut _) };
+        let inner: &'static mut SyncStream<S> = unsafe { &mut *(self.inner.deref_mut() as *mut _) };
```

Comment:
Add a safety note on 1. why the lifetime needs to be extended and 2. why it won't dangle